pkg/archive/zip.go

```diff
@@ -68,40 +70,156 @@ func (e ZipEntry) WriteTo(zw *zip.Writer) error {
+type MutableZipReader struct {
+	*zip.Reader
+	File []*MutableZipFile
+}
+
+func NewMutableReader(zr *zip.Reader) MutableZipReader {
+	mr := MutableZipReader{Reader: zr}
+	for _, zf := range zr.File {
+		mr.File = append(mr.File, &MutableZipFile{File: zf, FileHeader: zf.FileHeader})
+	}
+	return mr
+}
+
+func (mr MutableZipReader) WriteTo(zw *zip.Writer) error {
+	for _, mf := range mr.File {
+		r, err := mf.Open()
 		if err != nil {
 			return err
 		}
-		b, err := io.ReadAll(r)
+		w, err := zw.CreateHeader(&mf.FileHeader)
 		if err != nil {
-			r.Close()
 			return err
 		}
-		if err := r.Close(); err != nil {
+		if _, err := io.Copy(w, r); err != nil {
 			return err
 		}
-		// TODO: Memory-intensive. We're buffering the full file in memory (again).
-		// One option would be to do two passes and only buffer what's necessary.
-		ents = append(ents, ZipEntry{&zip.FileHeader{Name: f.Name, Modified: time.UnixMilli(0)}, b})
 	}
-	sort.Slice(ents, func(i, j int) bool {
-		return ents[i].FileHeader.Name < ents[j].FileHeader.Name
-	})
-	for _, ent := range ents {
-		w, err := zw.CreateHeader(ent.FileHeader)
-		if err != nil {
-			return err
-		}
-		if _, err := io.Copy(w, bytes.NewReader(ent.Body)); err != nil {
-			return err
+	return nil
+}
+
+type ZipArchiveStabilizer struct {
+	Name string
+	Func func(*MutableZipReader)
+}
+
+type ZipEntryStabilizer struct {
+	Name string
+	Func func(*MutableZipFile)
+}
+
+var AllZipStabilizers []any = []any{
+	StableZipFileOrder,
+	StableZipModifiedTime,
+	StableZipCompression,
+	StableZipDataDescriptor,
+	StableZipFileEncoding,
+	StableZipFileMode,
+	StableZipMisc,
+}
+
+var StableZipFileOrder = ZipArchiveStabilizer{
+	Name: "zip-file-order",
+	Func: func(zr *MutableZipReader) {
+		slices.SortFunc(zr.File, func(i, j *MutableZipFile) int {
+			return strings.Compare(i.Name, j.Name)
+		})
+	},
+}
+
+var StableZipModifiedTime = ZipEntryStabilizer{
+	Name: "zip-modified-time",
+	Func: func(zf *MutableZipFile) {
+		zf.Modified = time.UnixMilli(0)
+		zf.ModifiedDate = 0
+		zf.ModifiedTime = 0
+	},
+}
+
+var StableZipCompression = ZipEntryStabilizer{
+	Name: "zip-compression",
+	Func: func(zf *MutableZipFile) {
+		zf.Method = zip.Store
+	},
+}
+
+var StableZipDataDescriptor = ZipEntryStabilizer{
+	Name: "zip-data-descriptor",
+	Func: func(zf *MutableZipFile) {
+		zf.Flags = zf.Flags & ^uint16(0x8)
+		zf.CRC32 = 0
+		zf.CompressedSize = 0
+		zf.CompressedSize64 = 0
+		zf.UncompressedSize = 0
+		zf.UncompressedSize64 = 0
+	},
+}
+
+var StableZipFileEncoding = ZipEntryStabilizer{
+	Name: "zip-file-encoding",
+	Func: func(zf *MutableZipFile) {
+		zf.NonUTF8 = false
+	},
+}
+
+var StableZipFileMode = ZipEntryStabilizer{
+	Name: "zip-file-mode",
+	Func: func(zf *MutableZipFile) {
+		zf.CreatorVersion = 0
+		zf.ExternalAttrs = 0
+	},
+}
+
+var StableZipMisc = ZipEntryStabilizer{
+	Name: "zip-misc",
+	Func: func(zf *MutableZipFile) {
+		zf.Comment = ""
+		zf.ReaderVersion = 0
+		zf.Flags = zf.Flags & uint16(0x8)
```

Comment:
Related to the similar line in StableZipDataDescriptor?

Reply:
Yeah this just blanks out all flags except the data descriptor so they don't overlap in their functionality.

Reply:
It's probably worth adding that as a comment "Preserve the data descriptor flag to avoid overlap with zip-data-descriptor" or something?

Reply:
Done